### ISSUE TITLE
Allow tracing crates other than the current crate

### ISSUE THREAD

**Issue by reem** [dbrgn/tracing-test#6]
Due to the hardcoded env_filter here https://github.com/dbrgn/tracing-test/blob/main/tracing-test-macro/src/lib.rs#L67 it's only possible to trace logs from within the same crate as the test - would be nice to be able to specify additional crates/modules to include if the logs you're looking for are from dependency crates.

**Comment by reem** [comment]
Took a quick stab at it here but ran into the env filter actually being global in the current setup, so you can't set different ones for different tests, hopefully this is still a helpful start https://github.com/reem/tracing-test/tree/add-configurable-env-filter

**Comment by threema-danilo** [comment]
Hm, yeah, this would be nice to have. However, I don't know how to get around that `set_global_default` approach if we want to support logs from background threads.

Maybe as a workaround a `no-env-filter` cargo feature could be added, which disables the filtering globally?

**Comment by dvdsk** [comment]
Regarding the `set_global_default` did you take a look at [with_test_writer()](https://docs.rs/tracing-subscriber/0.3.1/tracing_subscriber/fmt/struct.SubscriberBuilder.html#method.with_test_writer)? I might misunderstand it and what this crate does but it allows me to set a custom filter function that I use to disallow/allow modules and crates. The procedural macro could generate the filter function from some attributes. I did not fully understand the issue with threads in the README so I might be completely off here.

```rust
fn setup_tracing() {
    use tracing_subscriber::{filter, prelude::*};

    let filter_modules = filter::filter_fn(|metadata| {
        if let Some(module) = metadata.module_path() {
            !module.contains("tarp")
        } else {
            true
        }
    });
    let fmt = tracing_subscriber::fmt::layer()
        .pretty()
        .with_test_writer();

    let _ignore_err = tracing_subscriber::registry()
        .with(fmt)
        .with(filter::LevelFilter::INFO)
        .with(filter_modules)
        .try_init();
}

```

**Comment by dbrgn** [comment]
@dvdsk IIRC the subscriber registration is thread-local, meaning that if you spawn a task that runs on another thread (or if you log from a classic thread), the logs will not be captured. I don't think this has changed lately, right?

**Comment by ijc** [comment]
I've just tripped over this.

> Maybe as a workaround a no-env-filter cargo feature could be added, which disables the filtering globally?

FWIW I'd be happy to opt in to that approach for my integration test crate.

Alternatively could `logs_with_scope_contain` be used to address any further filtering that a test needs, e.g. capture everything via the filter but then have `logs_assert` take a scope (provided by the macro generated wrapper function) but provide another API which asserts over all logs?


I also noticed that [`tracing_fmt::Builder`]https://docs.rs/tracing-fmt/0.0.1-alpha.3/tracing_fmt/struct.Builder.html) has methods ` with_filter_reloading`, `reload_handler` and `with_filter` etc which might allow something more dynamic compared with the `env_filter` approach.

**Comment by dbrgn** [comment]
> Alternatively could `logs_with_scope_contain` be used to address any further filtering that a test needs, e.g. capture everything via the filter but then have `logs_assert` take a scope (provided by the macro generated wrapper function) but provide another API which asserts over all logs?

Hmm, that could work! Would you be willing to give it a try?

(Please also check out the discussion in #12.)

> I also noticed that [`tracing_fmt::Builder`](https://docs.rs/tracing-fmt/0.0.1-alpha.3/tracing_fmt/struct.Builder.html) has methods ` with_filter_reloading`, `reload_handler` and `with_filter` etc which might allow something more dynamic compared with the `env_filter` approach.

With a global subscriber, that will only work if your tests do not run in parallel. Since `cargo test` runs tests in parallel by default, I wouldn't be happy about such a limitation.

**Comment by dvdsk** [comment]
> @dvdsk IIRC the subscriber registration is thread-local, meaning that if you spawn a task that runs on another thread (or if you log from a classic thread), the logs will not be captured. I don't think this has changed lately, right?

You are absolutely right, I ran into that :)

**Comment by rtraverso86** [comment]
I'm following this discussion, as the same issue prevents me to use tracing-test for integration tests within my same project as well. Since technically, each integration test is compiled as its own crate, most of the logs I'd want to keep and assert upon are filtered out.

**Comment by tami5** [comment]
> I'm following this discussion, as the same issue prevents me to use tracing-test for integration tests within my same project as well. Each integration test is compiled as its own crate, therefore most of the logs I'd want to keep and assert upon are filtered out.

+1

**Comment by threema-danilo** [comment]
An implementation of a `no-env-filter` feature is implemented in #16:

> ## Per-crate Filtering
>
> By default, `tracing-test` sets an env filter that filters out all logs
> except the ones from your crate (equivalent to
> `RUST_LOG=<your_crate>=trace`). If you need to capture logs from other crates
> as well, you can turn off this log filtering globally by enabling the
> `no-env-filter` Cargo feature:
>
> ```toml
> tracing-test = { version = "0.1", features = ["no-env-filter"] }
> ```
>
> Note that this will result in _all_ logs from _all_ your dependencies being
> captured! This means that the `logs_contain` function may become less
> useful, and you might need to use `logs_assert` instead, with your own
> custom filtering logic.

Could some of the people here please test this?

```diff
-tracing-test = "0.2.1"
+tracing-test = { git = "https://github.com/dbrgn/tracing-test/", branch = "no-env-filter", features = ["no-env-filter"] }
```
